src/Npgsql/Internal/TypeHandlers/ArrayHandler.cs

```diff
@@ -201,72 +196,91 @@ protected async ValueTask<List<TRequestedElement>> ReadList<TRequestedElement>(N
         internal static class ElementTypeInfo<TElement>
         {
             public static readonly bool IsNonNullable =
-                typeof(TElement).IsValueType && Nullable.GetUnderlyingType(typeof(TElement)) is null;
+                typeof(TElement).IsValueType && default(TElement) != null;
```

Comment:
Theoretically more important to use `is not` here, since TElement really could be any type (and `!=` may have been overridden).

```suggestion
                typeof(TElement).IsValueType && default(TElement) it not null;
```

Reply:
Good catch!